### DISCUSSION TITLE
Examples of a collection fixture?

### DISCUSSION THREAD

**Discussion by jplumhoff** [testcontainers/testcontainers-dotnet#412] [Q&A] [answered]
Hi,

I'm attempting to write xunit integration tests for repositories that store data in postgresql.  I'm using a collection fixture to create my postgresql container so it can be shared among all my repo test classes.  The tests all work perfectly when I step through them in debug, but every one of them fails if I just run them.  Each time, I get the same error for every test method:

```
Message: 
Npgsql.NpgsqlException : Exception while reading from stream
---- System.IO.EndOfStreamException : Attempted to read past the end of the stream.

  Stack Trace: 
NpgsqlReadBuffer.<Ensure>g__EnsureLong|41_0(NpgsqlReadBuffer buffer, Int32 count, Boolean async, Boolean readingNotifications)
NpgsqlConnector.RawOpen(SslMode sslMode, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
NpgsqlConnector.<Open>g__OpenCore|193_1(NpgsqlConnector conn, SslMode sslMode, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
NpgsqlConnector.Open(NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
ConnectorPool.OpenNewConnector(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
ConnectorPool.<Get>g__RentAsync|31_0(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
NpgsqlConnection.<Open>g__OpenAsync|45_0(Boolean async, CancellationToken cancellationToken)
SqlMapper.QueryAsync[T](IDbConnection cnn, Type effectiveType, CommandDefinition command) line 417
MaterialRepository.FindAsync(IQuerySpecification`1 querySpec) line 68
MaterialRepositoryTests.DeleteComponentMaterial() line 71
<>c.<ThrowAsync>b__128_0(Object state)
----- Inner Stack Trace -----
NpgsqlReadBuffer.<Ensure>g__EnsureLong|41_0(NpgsqlReadBuffer buffer, Int32 count, Boolean async, Boolean readingNotifications)
```

My first thought was that I was inadvertently re-using a DBconnection, but I create a new one for each test method. Since the tests all work if I create the container manually (without using testcontainers), I figure I have to be doing something wrong in the testcontainer fixture setup.  Does anybody have an example of a collection fixture they use for multiple test classes that they could share?  Super-extra bonus gratitude points if it's for a database...

**Comment by HofmeisterAn** [comment, marked as answer]
Can you share some of your code (especially the `CollectionDefinition`)? I use collection fixtures without any issues for databases too. It looks like your connection is broken somehow.

> The tests all work perfectly when I step through them in debug, but every one of them fails if I just run them.

Sounds like a race condition. Did you check the wait strategy? Is the container ready to accept connections, etc.?

**Edit**: I attached a small example. Maybe that helps: [PostgreSqlSample.zip](https://github.com/HofmeisterAn/dotnet-testcontainers/files/7750649/PostgreSqlSample.zip)

**Reply by jplumhoff** [reply]
Thanks for the example!  But as it turns out, your comment proved absolutely correct - the container wasn't ready to accept connections.  I had a WaitStrategy watching the logs for "database system is ready to accept connections".  Apparently, that's not good enough.  After seeing your response I changed it to the one you have in your postgres container:
`Wait.ForUnixContainer().UntilCommandIsCompleted($"pg_isready -h 'localhost' -p '5432'")`

Now everything is running perfectly.  Thanks for the pointer!  And just in general, thank you for this library - being able to run true integration tests against containers like this is a game changer...
